hera_cal/noise.py

```diff
@@ -0,0 +1,120 @@
+        kernel: differencing kernel for how to weight each visibiltiy relative to its neighbors
+            in time and frequency. Must sum to zero and must be 2D.
+
+    Returns:
+        variance: estimate of the noise variance on the input visibility with the same shape
+    '''
+    assert(np.sum(kernel) == 0), 'The kernal must sum to zero for difference-based noise estimation.'
+    assert(np.array(kernel).ndim == 2), 'The kernel must be 2D.'
+    variance = np.abs(scipy.signal.convolve2d(vis, kernel, mode='same', boundary='wrap'))**2
+    variance /= np.sum(np.array(kernel)**2)
+    return variance
+
+
+def predict_noise_variance_from_autos(bl, data, dt=None, df=None):
+    '''Predict the noise variance on a baseline using autocorrelation data.
+
+    Arguments:
+        bl: baseline tuple of the form (0, 1, 'xx')
+        data: DataContainer containing autocorrelation data
+        dt: integration time in seconds. If None, will try infer this
+            from the times stored in the DataContainer.
+        df: channel width in Hz. If None, will try to infer this from
+            from the frequencies stored in the DataContainer
+
+    Returns:
+        Noise variance predicted on baseline bl in units of data squared
+    '''
+    if dt is None:
+        assert(len(data.times_by_bl[bl[0:2]]) > 1)  # cannot infer integration time if only one integration is given
+        dt = np.median(np.ediff1d(data.times_by_bl[bl[0:2]])) * 24. * 3600.
+    if df is None:
+        assert(len(data.freqs) > 1)  # cannot infer channel width if only one channel is present
+        df = np.median(np.ediff1d(data.freqs))
+
+    ap1, ap2 = split_pol(bl[2])
+    return np.abs(data[bl[0], bl[0], join_pol(ap1, ap1)] * data[bl[1], bl[1], join_pol(ap2, ap2)] / dt / df)
+
+
+def per_antenna_noise_std(autos, dt=None, df=None):
+    '''Predicts per-antenna noise using autocorrelation data. The result is a noise standard deviation. To predict
+    the noise variance on e.g. (1, 2, 'xy'), one would use noise[1, 1, 'xx'] * noise[2, 2, 'yy'].
+    
+    Arguments:
+        autos: DataContainer containing autocorrelation data
+        dt: integration time in seconds. If None, will try infer this from the times stored in the DataContainer
+        df: channel width in Hz. If None, will try to infer this from the frequencies stored in the DataContainer
+
+    Returns:
+        noise: DataContainer mapping autocorrelation baselines like (1, 1, 'xx') to predictions for the standard
+            deviation of noise for that antenna.
+    '''
+    noise = {}
+    for bl in autos:
+        if (bl[0] == bl[1]) and (split_pol(bl[2])[0] == split_pol(bl[2])[1]):
+            noise[bl] = np.sqrt(predict_noise_variance_from_autos(bl, autos, dt=dt, df=df))
+    return DataContainer(noise)
+
+
+def write_per_antenna_noise_std_from_autos(infile, outfile, calfile=None, gain_convention='divide', add_to_history='', clobber=False):
```

Comment:
Despite the fact that this potentially replaces data with simulation, the default action is to not add any history remarks.  I'd prefer to see a default history message appended.

Reply:
Same for the rest of the functions which generate files.

Reply:
Thus far, this is how we've been doing this throughout hera_cal and I'd like to remain consistent (we use the run scripts to add to the history rather than the functions themselves). We have #394 open to do a better job of histories and I think this is a broader issue we should revisit as a group. If you'd like to add to that issue or open a new one, go ahead. But I think that's separate from this PR and should be handled all at once in a consistent way. 

I'll add it to the agenda for Wednesday.